### PR TITLE
ViperMonkey Feature Additions and Bug Fixes

### PR DESCRIPTION
**Pull request by kirk-sayre-work**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,22 +1,58 @@
 ViperMonkey
 ===========
 
+This is a fork of Philippe Lagadec's (http://www.decalage.info) ViperMonkey.
```
Comment (decalage2) [inline README.md]: Please remove this mention so that I can merge the PR.

File: README.md [hunk 2]
```diff
@@ -1,22 +1,58 @@
...
 Office applications, and all the DLLs and ActiveX objects that can be called from VBA.
 - This open-source project is only developed on my scarce spare time, so do not expect
 miracles. Any help from you will be very appreciated!
 
+**oletools Version**
+
+ViperMonkey requires the most recent version of [oletools]
+(https://github.com/decalage2/oletools), not the version downloaded
+pip. Make sure to either install the most recent oletools version by
+running the oletools 'setup.py install' or by making sure the most
```
Comment (decalage2) [inline README.md]: Or even better, see https://github.com/decalage2/oletools/wiki/Install#how-to-install-the-latest-development-version for a pip one-liner.

File: README.md [hunk 3]
```diff
@@ -31,18 +67,42 @@ miracles. Any help from you will be very appreciated!
...
 
-
-
 Download and Install:
 ---------------------
 
 For now, there is no package on PyPI for automated installation. It must be done manually:
 
-- Download the archive from the repository: https://github.com/decalage2/ViperMonkey/archive/master.zip
+- Download the archive from the repository: https://github.com/kirk-sayre-work/ViperMonkey/archive/master.zip
```
Comment (decalage2) [inline README.md]: Please revert to the main repo URL.

File: README.md [hunk 4]
```diff
@@ -56,35 +116,19 @@ To parse and interpret VBA macros from a document, use the vmonkey script:
...
 ```
 
-
-Documentation:
---------------
-
-Winter is coming.
-
-
-How to Suggest Improvements, Report Issues or Contribute:
```
Comment (decalage2) [inline README.md]: Please revert this deletion, this section is needed.

File: vipermonkey/core/expressions.py [hunk 5]
```diff
@@ -369,22 +443,21 @@ def eval(self, context, params=None):
...
-                                 ("*", 2, opAssoc.LEFT, Multiplication),
-                                 ("/", 2, opAssoc.LEFT, Division),
-                                 ("\\", 2, opAssoc.LEFT, FloorDivision),
-                                 (CaselessKeyword("mod"), 2, opAssoc.RIGHT, Mod),
-                                 ("+", 2, opAssoc.LEFT, Sum),
-                                 ("-", 2, opAssoc.LEFT, Subtraction),
-                                 ("&", 2, opAssoc.LEFT, Concatenation),
-                                 # (CaselessKeyword("xor"), 2, opAssoc.LEFT),
-                             ])
+expression <<= (operatorPrecedence(expr_item,
```
Comment (decalage2) [inline vipermonkey/core/expressions.py]: according to pyparsing's doc, operatorPrecedence is deprecated, should be replaced by infixNotation.

File: vipermonkey/core/identifiers.py [hunk 6]
```diff
@@ -136,10 +136,13 @@
...
-TODO_identifier_or_object_attrib = Combine(NotAny(reserved_keywords)
-    + Optional(Optional(entity_name) + Literal('.')) + Optional(entity_name + Literal('.')) + entity_name)
+TODO_identifier_or_object_attrib = Combine(NotAny(reserved_keywords) +
+                                           Optional(Optional(entity_name) + Literal('.')) +
+                                           Optional(entity_name + Literal('.')) +
+                                           entity_name +
+                                           # Looks like variables can end in $
+                                           Optional(CaselessLiteral('$')) +
+                                           # Looks like variables can end in '()'.
+                                           Suppress(Optional(CaselessLiteral('()'))))
```
Comment (decalage2) [inline vipermonkey/core/identifiers.py]: Not sure about that one... does not seem to be the best place to remove parentheses. What if it's a function call?

File: README.md (hunk 1 above)
Comment (kirk-sayre-work) [reply README.md]: Fixed in commit c963b624223cefed44fbb08bf7e70bbca0a6bf54.

File: README.md (hunk 3 above)
Comment (kirk-sayre-work) [reply README.md]: Fixed in commit c963b624223cefed44fbb08bf7e70bbca0a6bf54.

File: README.md (hunk 4 above)
Comment (kirk-sayre-work) [reply README.md]: Fixed in commit c963b624223cefed44fbb08bf7e70bbca0a6bf54.

File: vipermonkey/core/expressions.py (hunk 5 above)
Comment (kirk-sayre-work) [reply vipermonkey/core/expressions.py]: Fixed in commit c963b624223cefed44fbb08bf7e70bbca0a6bf54.

File: vipermonkey/core/identifiers.py (hunk 6 above)
Comment (kirk-sayre-work) [reply vipermonkey/core/identifiers.py]: I was having trouble parsing certain maldocs a few months ago that from what I remember had assignment statements like 'foo() = 12'. I have removed the '()' from the above parse expression in the current version as the recent maldocs I have been working and a simple test macro all with parse/emulate the same regardless of whether the '()' is included or not in this parse expression.

This change is in commit c963b624223cefed44fbb08bf7e70bbca0a6bf54.

File: vipermonkey/core/lib_functions.py [hunk 7]
```diff
@@ -109,8 +117,8 @@ def __repr__(self):
 # TODO: fix this like in olevba
 # chr_ = Suppress(Combine(CaselessLiteral('Chr') + Optional(Word('BbWw', max=1)) + Optional('$')) + '(') + expression + Suppress(')')
 chr_ = Suppress(Combine(
-    (CaselessKeyword('Chr') | CaselessKeyword('ChrB') | CaselessKeyword('ChrW'))
-    + Optional('$')) + '(') + expression + Suppress(')')
+                  (CaselessKeyword('Chr$') | CaselessKeyword('Chr') | CaselessKeyword('ChrB') | CaselessKeyword('ChrW'))
```
[the remark this answers is not in the document]
Comment (kirk-sayre-work) [reply vipermonkey/core/lib_functions.py]: The pyparse CaselessKeyword matches a literal string up to the first whitespace, punctuation, or other non-keyword character. It looks like pyparse does not include '$' as a non-keyword character, so parsing does not work correctly when doing CaselessKeyword('Chr') + Optional('$'). It works as intended if the '$' goes in the CaselessKeyword literal string being matched. I have added 'ChrB$' and 'ChrW$' in addition to 'Chr$' in the keywords being checked.

This is in commit c963b624223cefed44fbb08bf7e70bbca0a6bf54.

File: vipermonkey/core/identifiers.py (hunk 6 above)
Comment (decalage2) [reply vipermonkey/core/identifiers.py]: OK thanks.

**Pull request merged**